functions/src/replication/import.ts

```diff
@@ -0,0 +1,107 @@
+import * as admin from 'firebase-admin'
+import {
+  CollectionReference,
+  CollectionGroup,
+  QueryDocumentSnapshot,
+} from 'firebase-admin/firestore'
+import { SupabaseClient } from '@supabase/supabase-js'
+
+import { createSupabaseClient, run } from './utils'
+import { log, processPartitioned } from '../utils'
+import { initAdmin } from '../scripts/script-init'
+import { DocumentKind } from '../../../common/transaction-log'
+
+// strategy for live importing collection C without dropping data (times are firestore server times)
+// 1. optional - clear supabase table for collection C
+// 2. start replication of new collection C transaction log entries to supabase starting at T0
+// 3. establish import starting timestamp T1 such that T0 <= T1 <= now
+// 4. read collection C from firestore with effective timestamp T1 and send supabase writes
+// 5. supabase should be caught up
+
+async function getServerTimestamp() {
+  // firestore doesn't seem to have an API to just ask for the time, so we do this kludge
+  const result = await admin.firestore().collection('temp').doc('time').get()
+  return result.readTime.toDate()
+}
+
+function getWriteRow(
+  snap: QueryDocumentSnapshot,
+  docKind: DocumentKind,
+  ts: Date
+) {
+  return {
+    doc_kind: docKind,
+    write_kind: 'create',
+    doc_id: snap.id,
+    data: snap.data(),
+    ts: ts.toISOString(),
```

Comment:
So, timestamps are strings? Not milliseconds like we were using in firestore?

Reply:
In SQL there are specific date/time types that support date/time operations (see here: https://www.postgresql.org/docs/current/datatype-datetime.html) and reading/writing them is conventionally via ISO strings.